core/src/test/java/org/apache/iceberg/rest/responses/TestFetchScanTasksResponseParser.java

```diff
@@ -0,0 +1,180 @@
+  }
+
+  @Test
+  public void roundTripSerdeWithDeleteFilesNoFileScanTasksPresent() {
+    assertThatThrownBy(
+            () ->
+                FetchScanTasksResponse.builder()
+                    .withPlanTasks(List.of("task1", "task2"))
+                    .withDeleteFiles(List.of(FILE_A_DELETES))
+                    .build())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage(
+            "Invalid response: deleteFiles should only be returned with fileScanTasks that reference them");
+
+    String invalidJson =
+        "{\"plan-tasks\":[\"task1\",\"task2\"],"
+            + "\"delete-files\":[{\"spec-id\":0,\"content\":\"POSITION_DELETES\","
+            + "\"file-path\":\"/path/to/data-a-deletes.parquet\",\"file-format\":\"PARQUET\","
+            + "\"partition\":{\"1000\":0},\"file-size-in-bytes\":10,\"record-count\":1}]"
+            + "}";
+
+    assertThatThrownBy(() -> FetchScanTasksResponseParser.fromJson(invalidJson))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage(
+            "Invalid response: deleteFiles should only be returned with fileScanTasks that reference them");
+  }
+
+  @Test
+  public void roundTripSerdeWithFileScanTasks() {
+    ResidualEvaluator residualEvaluator =
+        ResidualEvaluator.of(SPEC, Expressions.equal("id", 1), true);
+    FileScanTask fileScanTask =
+        new BaseFileScanTask(
+            FILE_A,
+            new DeleteFile[] {FILE_A_DELETES},
+            SchemaParser.toJson(SCHEMA),
+            PartitionSpecParser.toJson(SPEC),
+            residualEvaluator);
+
+    FetchScanTasksResponse response =
+        FetchScanTasksResponse.builder()
+            .withFileScanTasks(List.of(fileScanTask))
+            .withDeleteFiles(List.of(FILE_A_DELETES))
+            // assume you have set this already
+            .withSpecsById(PARTITION_SPECS_BY_ID)
+            .build();
+
+    String expectedToJson =
+        "{"
+            + "\"delete-files\":[{\"spec-id\":0,\"content\":\"POSITION_DELETES\","
+            + "\"file-path\":\"/path/to/data-a-deletes.parquet\",\"file-format\":\"PARQUET\","
+            + "\"partition\":{\"1000\":0},\"file-size-in-bytes\":10,\"record-count\":1}],"
+            + "\"file-scan-tasks\":["
+            + "{\"data-file\":{\"spec-id\":0,\"content\":\"DATA\",\"file-path\":\"/path/to/data-a.parquet\","
+            + "\"file-format\":\"PARQUET\",\"partition\":{\"1000\":0},"
+            + "\"file-size-in-bytes\":10,\"record-count\":1,\"sort-order-id\":0},"
+            + "\"delete-file-references\":[0],"
+            + "\"residual-filter\":{\"type\":\"eq\",\"term\":\"id\",\"value\":1}}]"
+            + "}";
+
+    String json = FetchScanTasksResponseParser.toJson(response, false);
+    assertThat(json).isEqualTo(expectedToJson);
+
+    // make an unbound json where you expect to not have partitions for the data file,
+    // delete files as service does not send parition spec
```

Comment:
Generally try and avoid the second person in comments, "create a response where the file scan tasks are unbound" . It's not entirely clear or accurate to say "not have partitions for the data file", it's more so that there's no partitions that are in the deserialized version until it's bound.

Reply:
Will fix this